LLama/Model/IModelCache.cs

```diff
@@ -0,0 +1,45 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using LLama.Common;
+
+namespace LLama.Model;
+
+/// <summary>
+/// A class that helps organize and load local models
+/// </summary>
+public interface IModelCache : IDisposable
+{
+    /// <summary>
+    /// The current number of file handles in cache.
+    /// </summary>
+    /// <returns>Number of cached models</returns>
+    public int ModelsCached();
+
+    /// <summary>
+    /// Load a model file to be used for inference.
+    /// The caller assumes responsibility for disposing this model and <b>MUST</b> call Unload
+    /// </summary>
+    /// <param name="metadata"></param>
+    /// <param name="modelConfigurator"></param>
+    /// <param name="modelId">An alias to uniquely identify this model's underlying handle. If none is supplied, the model's name is used.'</param>
+    /// <param name="cancellationToken"></param>
+    /// <returns>The loaded model on success</returns>
+    public Task<LLamaWeights> LoadModelAsync(ModelFileMetadata metadata,
+        Action<ModelParams>? modelConfigurator = null!,
+        string modelId = "",
+        CancellationToken cancellationToken = default);
+
+    /// <summary>
+    /// Unload and dispose of a model with the given id
+    /// </summary>
+    /// <param name="modelId"></param>
+    /// <returns></returns>
+    public bool UnloadModel(string modelId);
+
+    /// <summary>
+    /// Unload all currently loaded models
+    /// </summary>
+    public void UnloadAllModels();

```

Comment:
There's a few challenges with supporting this and I'm thinking maybe it's better to just drop these methods and have the caller explicitly call `Dispose` on the weight? 

A few questions
1. Should multiple instances of the same model be allowed to be loaded? I think yes. If so, we'll need a way to ensure calling unload is specific to a model.
2. Should we force model aliases in the cache to be unique? This would help the issue where I load multiple instances of the model and then calling dispose is guaranteed to operate on that same model. If we don't want to enforce this restriction, unloading the correct model because tricker and might require the original model to be passed in as opposed to the alias. 
3. Are we better off getting rid of this class altogether? My main goal was to have something like `IModelRepo` and quickly went out of scope with some of this but if we think it'll be useful I'm happy to leave it but I'm struggling a bit to justify it especially given the unload/dispose challenges.

Reply:
> Should multiple instances of the same model be allowed to be loaded? I think yes. If so, we'll need a way to ensure calling unload is specific to a model.

Definitely. You can load a model with different settings (e.g. loras) which affect the output almost as if it's another model.

> Should we force model aliases in the cache to be unique?

Do you mean the `modelId` string? If so then I would think loading a model with an ID which is in the cache would either:
 - throw an error
 - return the existing model

I'd lean towards throwign the error out of those choices.

> Are we better off getting rid of this class altogether?

The model repo idea is interesting, and I think it's something Rinne has looked at as well. But yeah maybe it would be better to remove it from this PR, and make this PR all about the new shared lifetimes (which I think is a pretty big improvement in itself). The model repo can be done in a followup PR, fully taking advantage of the new lifetimes.

Reply:
If anything, I'd rather keep the model repo and the file system class than this but I've further reworked this to be more explicit. I haven't been able to finalize what a "good" api is for something like this (name aside) but the discussion here has helped add context.